validator/client/src/main/java/tech/pegasys/teku/validator/client/DoppelgangerDetectionService.java

```diff
@@ -0,0 +1,181 @@
+    if (duration.compareTo(timeout) > 0) {
+      LOGGER.info("Doppelganger Detection Service max allowed duration exceeded.");
+      return stop().thenRun(() -> doppelgangerCheckFinished.set(true));
+    }
+
+    return genesisDataProvider
+        .getGenesisTime()
+        .thenCompose(
+            genesisTime -> {
+              final UInt64 currentEpoch =
+                  spec.computeEpochAtSlot(
+                      spec.getCurrentSlot(timeProvider.getTimeInMillis(), genesisTime));
+
+              if (epochAtStart.isEmpty()) {
+                epochAtStart = Optional.of(currentEpoch);
+              }
+
+              if (currentEpoch.minus(epochAtStart.get()).isGreaterThan(2)) {
+                return stop().thenRun(() -> doppelgangerCheckFinished.set(true));
+              }
+
+              return validatorIndexProvider
+                  .getValidatorIndices()
+                  .thenCompose(
+                      validatorIndices ->
+                          validatorApiChannel
+                              .checkValidatorsDoppelganger(
+                                  validatorIndices
+                                      .intStream()
+                                      .mapToObj(UInt64::valueOf)
+                                      .collect(Collectors.toList()),
+                                  currentEpoch)
+                              .thenApply(
+                                  validatorLivenessAtEpoches ->
+                                      doppelgangerDetected(
+                                          validatorLivenessAtEpoches,
+                                          validatorIndices
+                                              .intStream()
+                                              .mapToObj(UInt64::valueOf)
+                                              .collect(Collectors.toList())))
+                              .exceptionally(
+                                  throwable -> {
+                                    LOGGER.error(
+                                        "Unable to check validators doppelganger: {}",
+                                        throwable.getMessage());
+                                    return false;
+                                  })
+                              .thenApply(
+                                  doppelgangerDetected -> {
+                                    if (doppelgangerDetected) {
+                                      LOGGER.fatal(
+                                          "Doppelganger detected. Shutting down Validator Client.");
+                                      System.exit(1);
+                                    }
+                                    return null;
+                                  }));
+            })
+        .exceptionally(
+            throwable -> {
+              LOGGER.error("Unable to check validators doppelganger: {}", throwable.getMessage());
```

Comment:
again, the user will read this message, they need to understand it and potentially take action...

Reply:
Done, added more details here